### PR TITLE
initial commit for backend-smda

### PR DESCRIPTION
**Pull request by danielplohmann**
Hey all!

We propose to integrate [SMDA](https://github.com/danielplohmann/smda) as a disassembler backend into capa to enable full Python 3 support.
SMDA is a lightweight and fast recursive disassembler built on top of [capstone](https://github.com/aquynh/capstone), covering PE and ELF files in 32/64bit. It is compatible with both Python 2 and Python 3.

The pull request contains an implementation structurally oriented along the vivisect implementation and provides all capa-features covered by vivisect. We also made sure that the code is in line with the style requirements and has full test coverage using the `tests/fixtures.get_extractor()` method:

CapaTests

This initial PR sets SMDA as a default backend for Python 3. 
This choice was made to allow you an easier evaluation of the contribution. 

As discussed out of band, we don't necessarily ask to make SMDA the default backend for Python 3 but instead propose to allow users choosing their preferred backend through a CLI parameter, which could also be used to pass additional parameters to the disassembler backend (e.g. SMDA can process shellcode / memory dumps and make use of [ApiScout](https://github.com/danielplohmann/apiscout) for reconstruction of dynamic imports - increasing the potential matches of rules).

This PR was a collaborative effort of @jcrussell and @danielplohmann during [GeekWeek 7](https://cyber.gc.ca/en/events/geekweek-7).

### CODE REVIEW

**Review by williballenthin — CHANGES_REQUESTED** [review]
this is a really great piece of work, nice job.

I can tell from the changes you made that you've gotten this working quite successfully. I also appreciate that the style is consistent, both within the PR and our existing code base - that really helps us integrate the changes a lot easier.

The biggest thing I'd like to see before merging is the addition of feature unit tests against the SMDA backend. there are 100 or so unit tests that ensure that each backend behaves essentially the same. fortunately, registering the SMDA backend should be pretty easy (if not, its a bug, and I'd like to work to make it easy for you).

would you duplicate [test_viv_features.py](https://github.com/fireeye/capa/blob/master/tests/test_viv_features.py) as `test_smda_features.py` and update it appropriately?

**Comment by williballenthin** [comment]
also, i broke master by merging a rule but not syncing the testfiles submodule. thats since fixed. if you can sync this branch from master then we should see the test results show up in CI.

File: capa/features/extractors/smda/insn.py [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+    Characteristic,
+)
+from capa.features.insn import API, Number, Offset, Mnemonic
+
+# security cookie checks may perform non-zeroing XORs, these are expected within a certain
+# byte range within the first and returning basic blocks, this helps to reduce FP features
+SECURITY_COOKIE_BYTES_DELTA = 0x40
+
+
+def get_arch(smda_report: SmdaReport):
```
[the remark this answers is not in the document]
Comment (danielplohmann) [reply capa/features/extractors/smda/insn.py]: yeah, I just recently started using them and I'm a big fan. :)

**Comment by danielplohmann** [comment]
> this is a really great piece of work, nice job.
> 
> I can tell from the changes you made that you've gotten this working quite successfully. I also appreciate that the style is consistent, both within the PR and our existing code base - that really helps us integrate the changes a lot easier.
> 
> The biggest thing I'd like to see before merging is the addition of feature unit tests against the SMDA backend. there are 100 or so unit tests that ensure that each backend behaves essentially the same. fortunately, registering the SMDA backend should be pretty easy (if not, its a bug, and I'd like to work to make it easy for you).
> 
> would you duplicate [test_viv_features.py](https://github.com/fireeye/capa/blob/master/tests/test_viv_features.py) as `test_smda_features.py` and update it appropriately?

20201030-capa-smda


Alright, here we go!
Fully implemented and (almost) passing tests for `test_smda_features.py`.

We found that the only failing test (x64 nested thunks) does not succeed because disassemblers disagree where to define function borders. It seems vivisect wants `0x140059342` as function start, while IDA/SMDA want `0x14005E0C0` as a function start and say  `0x140059342` is just a function with a single jump instruction to `0x14005E0C0`. 
In `fixtures.py`, I commented on this and also proposed another, simpler function `0x1400615c0` which would serve as possible replacement for a x64 nested thunks test. It's calling `IsProcessorFeaturePresent` via double indirection, but this function appears to be missed by vivisect. :(
Please advise how to proceed.

File: capa/features/extractors/smda/function.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
+from capa.features import Characteristic
+from capa.features.extractors import loops
+
+
+def interface_extract_function_XXX(f):
```
Comment (mr-tz) [inline capa/features/extractors/smda/function.py:5]: can be removed

File: capa/main.py [hunk 3]
```diff
@@ -446,14 +458,25 @@ def main(argv=None):
...
-        help="path to sample to analyze",
-    )
+    # TODO: decode won't work for python3
+    if sys.version_info >= (3, 0):
+        parser.add_argument(
+            # in #328 we noticed that the sample path is not handled correctly if it contains non-ASCII characters
+            # https://stackoverflow.com/a/22947334/ offers a solution and decoding using getfilesystemencoding works
+            # in our testing, however other sources suggest `sys.stdin.encoding` (https://stackoverflow.com/q/4012571/)
+            "sample",
+            type=str,
```
[the remark this answers is not in the document]
Comment (mr-tz) [reply capa/main.py:468]: great

**Review by mr-tz — COMMENTED** [review]
very cool, this looks great, I think the following items are left to do:
- find a new nested thunk test case that works across the board (#356)
- decide how to use/specify the analysis backend
  - specify command-line argument(s)
  - default to vivisect for Py2 and SMDA for Py3?

**Review by williballenthin — APPROVED** [review]
lgtm

@mr-tz  @mike-hunhoff  will one of you take a second look and then do the merge?

File: capa/features/extractors/smda/function.py (hunk 2 above)
Comment (mike-hunhoff) [reply capa/features/extractors/smda/function.py:5]: agreed

File: capa/features/extractors/smda/insn.py [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+
+    args:
+      f (smda.common.SmdaFunction): the function to process.
+      bb (smda.common.SmdaBasicBlock): the basic block to process.
+      insn (smda.common.SmdaInstruction): the instruction to process.
+
+    yields:
+      (Feature, int): the feature and the address at which its found.
+    """
+    yield NotImplementedError("feature"), NotImplementedError("virtual address")
```
Comment (mike-hunhoff) [inline capa/features/extractors/smda/insn.py:48]: can be removed

**Review by mike-hunhoff — CHANGES_REQUESTED** [review]
Awesome, thank you for all of your hard work on this implementation! It's really cool to see capa running in Python 3 and your approach to implementing a new extractor for `capa`. I left a few comments to consider but overall this looks great!

I tested locally and encountered errors while running the `smda` tests on Windows with Python v3.7.8 (see below):

```
> C:\Exclusions\capa>python -m pytest tests\test_smda_features.py
...
=========================================================================================================== short test summary info ===========================================================================================================
FAILED tests/test_smda_features.py::test_smda_features[kernel32-64-function=0x180001010-api(RtlVirtualUnwind)-True0] - struct.error: unpack requires a buffer of 4 bytes
FAILED tests/test_smda_features.py::test_smda_features[kernel32-64-function=0x180001010-api(RtlVirtualUnwind)-True1] - struct.error: unpack requires a buffer of 4 bytes
FAILED tests/test_smda_features.py::test_smda_features[kernel32-64-function=0x1800202B0-api(RtlCaptureContext)-True0] - struct.error: unpack requires a buffer of 4 bytes
FAILED tests/test_smda_features.py::test_smda_features[kernel32-64-function=0x1800202B0-api(RtlCaptureContext)-True1] - struct.error: unpack requires a buffer of 4 bytes
FAILED tests/test_smda_features.py::test_smda_features[al-khaser x64-function=0x14004B4F0-api(__vcrt_GetModuleHandle)-True] - struct.error: unpack requires a buffer of 4 bytes
FAILED tests/test_smda_features.py::test_smda_features[kernel32-64-function=0x1800017D0-characteristic(peb access)-True] - struct.error: unpack requires a buffer of 4 bytes
FAILED tests/test_smda_features.py::test_smda_features[kernel32-64-function=0x180001068-characteristic(gs access)-True] - struct.error: unpack requires a buffer of 4 bytes
FAILED tests/test_smda_features.py::test_smda_features[a1982...-function=0x4014D0-characteristic(cross section flow)-True] - struct.error: unpack requires a buffer of 4 bytes
FAILED tests/test_smda_features.py::test_smda_features[kernel32-64-function=0x180001068-characteristic(cross section flow)-False] - struct.error: unpack requires a buffer of 4 bytes

Results (47.59s):
     104 passed
       9 failed
         - tests/test_smda_features.py:13 test_smda_features[kernel32-64-function=0x180001010-api(RtlVirtualUnwind)-True0]
         - tests/test_smda_features.py:13 test_smda_features[kernel32-64-function=0x180001010-api(RtlVirtualUnwind)-True1]
         - tests/test_smda_features.py:13 test_smda_features[kernel32-64-function=0x1800202B0-api(RtlCaptureContext)-True0]
         - tests/test_smda_features.py:13 test_smda_features[kernel32-64-function=0x1800202B0-api(RtlCaptureContext)-True1]
         - tests/test_smda_features.py:13 test_smda_features[al-khaser x64-function=0x14004B4F0-api(__vcrt_GetModuleHandle)-True]
         - tests/test_smda_features.py:13 test_smda_features[kernel32-64-function=0x1800017D0-characteristic(peb access)-True]
         - tests/test_smda_features.py:13 test_smda_features[kernel32-64-function=0x180001068-characteristic(gs access)-True]
         - tests/test_smda_features.py:13 test_smda_features[a1982...-function=0x4014D0-characteristic(cross section flow)-True]
         - tests/test_smda_features.py:13 test_smda_features[kernel32-64-function=0x180001068-characteristic(cross section flow)-False]
```

The failed tests appear to be the result of the following error:

```
C:\Exclusions\capa>python -m capa.main tests\data\al-khaser_x64.exe_
WARNING:capa:skipping non-.yml file: 11736.py
loading : 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 422/422 [00:00<00:00, 1227.46     rules/s]
Traceback (most recent call last):
  File "C:\Program Files\Python37\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "C:\Program Files\Python37\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "C:\Exclusions\capa\capa\main.py", line 713, in <module>
    sys.exit(main())
  File "C:\Exclusions\capa\capa\main.py", line 595, in main
    extractor = get_extractor(args.sample, args.format, disable_progress=args.quiet)
  File "C:\Exclusions\capa\capa\main.py", line 319, in get_extractor
    return get_extractor_py3(path, format, disable_progress=disable_progress)
  File "C:\Exclusions\capa\capa\main.py", line 308, in get_extractor_py3
    smda_report = smda_disasm.disassembleFile(path)
  File "smda\Disassembler.py", line 37, in disassembleFile
    loader = FileLoader(file_path, map_file=True)
  File "smda\utility\FileLoader.py", line 17, in __init__
    self._loadFile()
  File "smda\utility\FileLoader.py", line 32, in _loadFile
    self._base_addr = loader.getBaseAddress(self._raw_data)
  File "smda\utility\PeFileLoader.py", line 94, in getBaseAddress
    base_addr = struct.unpack("L", binary[pe_offset + 0x30:pe_offset + 0x38])[0]
struct.error: unpack requires a buffer of 4 bytes
```

Let's address this error and then we should be ready for a merge 🚀

File: capa/features/extractors/smda/basicblock.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+        return False
+
+    return True
+
+
+def is_printable_ascii(chars):
+    if sys.version_info[0] >= 3:
+        return all(c < 127 and chr(c) in string.printable for c in chars)
+    else:
+        return all(ord(c) < 127 and c in string.printable for c in chars)
```
[the remark this answers is not in the document]
Comment (williballenthin) [reply capa/features/extractors/smda/basicblock.py:75]: should we have an assert somewhere to codify that smda will only run on py3?

**Comment by danielplohmann** [comment]
> Awesome, thank you for all of your hard work on this implementation! It's really cool to see capa running in Python 3 and your approach to implementing a new extractor for `capa`. I left a few comments to consider but overall this looks great!
> 
> I tested locally and encountered errors while running the `smda` tests on Windows with Python v3.7.8 (see below):
> 
> ```
> > C:\Exclusions\capa>python -m pytest tests\test_smda_features.py
> ...
> =========================================================================================================== short test summary info ===========================================================================================================
> FAILED tests/test_smda_features.py::test_smda_features[kernel32-64-function=0x180001010-api(RtlVirtualUnwind)-True0] - struct.error: unpack requires a buffer of 4 bytes
> FAILED tests/test_smda_features.py::test_smda_features[kernel32-64-function=0x180001010-api(RtlVirtualUnwind)-True1] - struct.error: unpack requires a buffer of 4 bytes
> FAILED tests/test_smda_features.py::test_smda_features[kernel32-64-function=0x1800202B0-api(RtlCaptureContext)-True0] - struct.error: unpack requires a buffer of 4 bytes
> FAILED tests/test_smda_features.py::test_smda_features[kernel32-64-function=0x1800202B0-api(RtlCaptureContext)-True1] - struct.error: unpack requires a buffer of 4 bytes
> FAILED tests/test_smda_features.py::test_smda_features[al-khaser x64-function=0x14004B4F0-api(__vcrt_GetModuleHandle)-True] - struct.error: unpack requires a buffer of 4 bytes
> FAILED tests/test_smda_features.py::test_smda_features[kernel32-64-function=0x1800017D0-characteristic(peb access)-True] - struct.error: unpack requires a buffer of 4 bytes
> FAILED tests/test_smda_features.py::test_smda_features[kernel32-64-function=0x180001068-characteristic(gs access)-True] - struct.error: unpack requires a buffer of 4 bytes
> FAILED tests/test_smda_features.py::test_smda_features[a1982...-function=0x4014D0-characteristic(cross section flow)-True] - struct.error: unpack requires a buffer of 4 bytes
> FAILED tests/test_smda_features.py::test_smda_features[kernel32-64-function=0x180001068-characteristic(cross section flow)-False] - struct.error: unpack requires a buffer of 4 bytes
> 
> Results (47.59s):
>      104 passed
>        9 failed
>          - tests/test_smda_features.py:13 test_smda_features[kernel32-64-function=0x180001010-api(RtlVirtualUnwind)-True0]
>          - tests/test_smda_features.py:13 test_smda_features[kernel32-64-function=0x180001010-api(RtlVirtualUnwind)-True1]
>          - tests/test_smda_features.py:13 test_smda_features[kernel32-64-function=0x1800202B0-api(RtlCaptureContext)-True0]
>          - tests/test_smda_features.py:13 test_smda_features[kernel32-64-function=0x1800202B0-api(RtlCaptureContext)-True1]
>          - tests/test_smda_features.py:13 test_smda_features[al-khaser x64-function=0x14004B4F0-api(__vcrt_GetModuleHandle)-True]
>          - tests/test_smda_features.py:13 test_smda_features[kernel32-64-function=0x1800017D0-characteristic(peb access)-True]
>          - tests/test_smda_features.py:13 test_smda_features[kernel32-64-function=0x180001068-characteristic(gs access)-True]
>          - tests/test_smda_features.py:13 test_smda_features[a1982...-function=0x4014D0-characteristic(cross section flow)-True]
>          - tests/test_smda_features.py:13 test_smda_features[kernel32-64-function=0x180001068-characteristic(cross section flow)-False]
> ```
> 
> The failed tests appear to be the result of the following error:
> 
> ```
> C:\Exclusions\capa>python -m capa.main tests\data\al-khaser_x64.exe_
> WARNING:capa:skipping non-.yml file: 11736.py
> loading : 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 422/422 [00:00<00:00, 1227.46     rules/s]
> Traceback (most recent call last):
>   File "C:\Program Files\Python37\lib\runpy.py", line 193, in _run_module_as_main
>     "__main__", mod_spec)
>   File "C:\Program Files\Python37\lib\runpy.py", line 85, in _run_code
>     exec(code, run_globals)
>   File "C:\Exclusions\capa\capa\main.py", line 713, in <module>
>     sys.exit(main())
>   File "C:\Exclusions\capa\capa\main.py", line 595, in main
>     extractor = get_extractor(args.sample, args.format, disable_progress=args.quiet)
>   File "C:\Exclusions\capa\capa\main.py", line 319, in get_extractor
>     return get_extractor_py3(path, format, disable_progress=disable_progress)
>   File "C:\Exclusions\capa\capa\main.py", line 308, in get_extractor_py3
>     smda_report = smda_disasm.disassembleFile(path)
>   File "smda\Disassembler.py", line 37, in disassembleFile
>     loader = FileLoader(file_path, map_file=True)
>   File "smda\utility\FileLoader.py", line 17, in __init__
>     self._loadFile()
>   File "smda\utility\FileLoader.py", line 32, in _loadFile
>     self._base_addr = loader.getBaseAddress(self._raw_data)
>   File "smda\utility\PeFileLoader.py", line 94, in getBaseAddress
>     base_addr = struct.unpack("L", binary[pe_offset + 0x30:pe_offset + 0x38])[0]
> struct.error: unpack requires a buffer of 4 bytes
> ```
> 
> Let's address this error and then we should be ready for a merge

Hehe, seems like I fell into the struct format char size trap - I now changed this to `Q`, so it should be fine across all platforms.
All other comments are addressed.

What's still open for discussion:
1. find a new nested thunk test case that works across the board (#356)
1. decide how to use/specify the analysis backend
   1. specify command-line argument(s)
   1.  default to vivisect for Py2 and SMDA for Py3?

1) I just checked this in my python2 venv and it looks my proposed test replacement for a x64 nested thunk is processed correctly by vivisect as well - if you can confirm this, we would be able to close #356 as well.
2.i) I would go ahead in the near future and prepare this in a separate PR.
2.ii) That is at least the case for now. As stated initially, feel free to change out SMDA as default any time if another python3 disassembler makes a good replacement. 2.i) would then at least conveniently allow for SMDA to remain available.

**Comment by mr-tz** [comment]
1. Thanks, the test case is great! This closes #356 
2. That works for me.

Before we merge, can you please:
- Add a check to verify Python3, see https://github.com/fireeye/capa/pull/355#discussion_r518366831; ref: https://stackoverflow.com/questions/62569556/how-to-graceful-exit-if-the-required-verion-of-python-is-not-used
- Investigate why the Python 3.9.0-rc.1 setup fails, see https://github.com/fireeye/capa/pull/355/checks?check_run_id=1374027450#step:5:249
  - Do we need a workaround?
- Merge in the improvements I suggest in https://github.com/fireeye/capa/compare/mrtz-backend-smda, see https://github.com/fireeye/capa/pull/360

**Comment by danielplohmann** [comment]
* @mr-tz I've just merged your improvements, thank you!
 * I've put the runtime check for Python 3 into the [initialization](https://github.com/fireeye/capa/pull/355/commits/f7492c7dc7be9fe526fb5c8d0dffc87a8be6af9c#diff-0f8fe24da6e17fcaa0cbef5eceaeabec108167c53f34470b0175ec5718b4f40fR18) of SmdaFeatureExtractor and throw a `UnsupportedRuntimeError` in case this condition is not met. Is this sufficient or would anyone prefer a different way of handling this?
 * The failure of Python 3.9.0-rc.1 is pretty unfortunate as this is due to LIEF not being available via PyPI for Python 3.9 yet. There are recent commits in their repo working towards supporting it. It seems that it is also already possible to compile it from source for Python 3.9 using the Github repository, so I assume it will be available for Python 3.9 in the near future.

**Comment by mr-tz** [comment]
Great, this looks good to me then. Well done and many thanks again!
@williballenthin, @mike-hunhoff are we good to merge?

**Comment by williballenthin** [comment]
lets do it!

**Pull request merged**